### ISSUE TITLE
Chip-type dependent includes and compilation

### ISSUE THREAD

**Issue by dbaumgarten** [dbaumgarten/yodk#50]
There should be a way to choose what kind of chip a nolol-script should be compiled for.
If the script uses operations not supported by the chip, compilation should fail.

When including files, it should be possible to have different files for different chip-types and have the correct on choosen automatically. e.g:
- file_basic.nolol
- file_advanced.nolol

"include "file" " will automatically choose the correct one, depending on the current compilation-target.

closely related to #49

**Comment by Woccz** [comment]
Can I also add that all files up to the chip level be included, but higher levels should override the methods of lower levels?

For example, in `file_basic.nolol`, there is the code
```
Macro a = x+y
Macro b = x/1000*1000
```
and in `file_advanced.nolol` there is 
```
Macro b = x-x%1
```

If compiled with the basic compiler it should give you the functions
```
Macro a = x+y
Macro b = x/1000*1000
```
and with the advanced compiler
```
Macro a = x+y
Macro b = x-x%1
```

Do you think this would be beneficial? It would eliminate the need to rewrite all macros for all chip types, and just override ones that have a better implementation.

Thank you.

**Comment by dbaumgarten** [comment]
I was thinking if it would be better to completely ditch the file-name-based approach and do something like ```#if``` in C.
Special if-blocks that can only have static expressions as condition and are evaluated at compile-time.
This way macros could simply be wrapped in compile-time-ifs that check the target chiptype.

```
#if TYPE=="basic" then
  macro floor(x) expr
    x/1000*1000
  end
#else
 macro floor(x) expr
    x-x%1
  end
end
```

That would reduce code-duplication to an absolute minimum and would even allow other things like compile-time-configurable scripts. The user just changes some definitions at the top of the file and based on this certain parts of the code are added or removed during compilation.

It would be quite powerfull, but there are a few issues.  
- It's ugly
- People would probably write horrible code using this (C preprocessor hell...)
- Currently includes and macro-definitions are only allowed on the top-level (not inside any blocks)
- Fits a little awkwardly into the current nolol-compiler implementation


The filename approach is less powerfull, but at least the code gets less messy.

Not really sure what the best thing to do is here.

**Comment by Woccz** [comment]
I see your idea, and it would help to keep everything in the same file, but it is definitely ugly. I'll think about finding a tidy way to do this... maybe @ tags, e.g. ```
macro floor(x) expr
    x/1000*1000
end

@advanced
macro floor(x) expr
    x-x%1
end
```
would look much tidier?

**Comment by dbaumgarten** [comment]
That's roughly how rust does it.
It would only work for macros, and not for pure code and definitions, but I guess that could be enough.

The nice thing with the file-name-based approach would be that it would automatically work nicely with the "is this function available"-check. It would work the same way.

To reduce code-duplications a programmer could just use includes.
Like add "include file_advanced" to the top of your file_professinal.nolol.  This way file_professional.nolol includes all the advanced code, but macro-definitions inside it Overwrite existing ones.

Basic Idea:
The compiler tries to include an exact match.
If these isn't on, it tries to include _professional, _advanced, _basic in that order, starting with the current Chip-Type.
If the programmer has code that is shared between these files, he puts it in a separate file and includes that in the other files.

**Comment by Woccz** [comment]
I basically did something lile this manually for my own library. "my_lib" included "my_lib_advanced", included "my_lib_basic".
It worked pretty well, but would be nice to not have to include the lower level librarys, and to have the compiler automatically select include level for all includes, rather than having to change it every one.

**Comment by dbaumgarten** [comment]
Making progress here. When compiling (or editing) nolol-scripts a chip-type can now be provided via cli-flag or vscode setting. The default is auto-detection, which uses the file-name o fthe compiled file to determine the chiptype.

Currently it only checks if the script uses any operations that are unavailable on the chosen chip and if so throws an error.

Macros are only check if they are either defined in the current file, or are actually used in the code.
(This way unused macros in included files don't throw errors, but when you are currently writing a macro you get notified about unavailable ops, even when the macros currently isn't used anywhere.

Next steps: Being able do specifiy chiptype for testcases and chiptype-specific includes.

**Comment by dbaumgarten** [comment]
Done. Published as part of v0.1.5
